src/index.ts

```diff
@@ -51,351 +41,292 @@ export = class ChunksWebpackPlugin {
 			baseDataPath: 'options'
 		});
 
-		this.manifest = {};
-		this.addAssets = this.addAssets.bind(this);
+		this.hookCallback = this.hookCallback.bind(this);
 	}
 
 	/**
 	 * Apply function is automatically called by the Webpack main compiler
 	 * @param {Object} compiler The Webpack compiler variable
 	 */
 	apply(compiler: Compiler): void {
-		compiler.hooks.thisCompilation.tap('ChunksWebpackPlugin', this.hookCallback.bind(this));
+		compiler.hooks.thisCompilation.tap('ChunksWebpackPlugin', this.hookCallback);
 	}
 
 	/**
 	 * Hook expose by the Webpack compiler
-	 * @param {Object} compilation The Webpack compilation variable
+	 * @async
+	 * @param {Compilation} compilation Webpack compilation
 	 */
-	hookCallback(compilation: object): void {
-		this.compilation = compilation;
-		this.fs = this.compilation.compiler.outputFileSystem;
-
-		const stage = this.options.outputPath
-			? Infinity
-			: webpack.Compilation.PROCESS_ASSETS_STAGE_ADDITIONAL;
-		this.compilation.hooks.processAssets.tap(
+	async hookCallback(compilation: Compilation): Promise<void> {
+		compilation.hooks.processAssets.tapPromise(
 			{
 				name: 'ChunksWebpackPlugin',
-				stage
+				stage: webpack.Compilation.PROCESS_ASSETS_STAGE_ADDITIONAL
 			},
-			this.addAssets
+			this.addAssets.bind(this, compilation)
 		);
 	}
 
 	/**
 	 * Add assets
 	 * The hook is triggered by webpack
+	 * @async
+	 * @param {Compilation} compilation Webpack compilation
+	 * @returns {Promise<void>}
 	 */
-	addAssets(): void {
-		this.publicPath = this.getPublicPath();
-		this.outputPath = this.getOutputPath();
-		this.pathFromFilename = this.getPathFromString(this.options.filename);
-		this.entryNames = this.getEntryNames();
-
-		this.entryNames
-			.filter((entryName: string) => this.getFiles(entryName).length)
-			.map((entryName: string) => this.processEntry(entryName));
-
-		// Check if manifest option is enabled
-		if (this.options.generateChunksManifest) {
-			this.createChunksManifestFile();
+	async addAssets(compilation: Compilation): Promise<void> {
+		// For better compatibility with future webpack versions
+		const RawSource = compilation.compiler.webpack.sources.RawSource;
+
+		const cache = compilation.getCache('ChunksWebpackPlugin');
+		const entryNames = compilation.entrypoints.keys();
+		const entryCssData: Array<EntryCssData> = [];
+		const entryJsData: Array<EntryJsData> = [];
+		const manifest: Manifest = {};
+
+		await Promise.all(
+			Array.from(entryNames).map(async (entryName: string) => {
+				const filesDependencies = this.getFilesDependenciesByEntrypoint({
+					compilation,
+					entryName
+				});
+
+				manifest[entryName] = {
+					styles: [],
+					scripts: []
+				};
+				const publicPath = this.getPublicPath(compilation);
+
+				if (filesDependencies.css.length) {
+					const eTagCss = filesDependencies.css.map((item: Asset) =>
```

Comment:
You need to emrge `eTagCss`, because there is an array, but you need only one eTag

Reply:
Like here https://github.com/yoriiis/svg-chunk-webpack-plugin/blob/main/src/index.ts#L136

Reply:
Using `reduce`

Reply:
Well seen ! It works as expected 👍🏻